src/helpers.php

```diff
@@ -38,88 +38,43 @@ function get_annotations(string $from, string $annotationName): array
 
-        // Checking the return value actually requires executing the method.
-        // So use this to avoid infinite recursion.
-        $currentMethod = collect(explode('::', __METHOD__))->last();
-        $filter = $types[$heritage];
         // The method must be public
         $methods = $reflectionClass->getMethods(ReflectionMethod::IS_PUBLIC);
 
         $methods = collect($methods)
-            ->filter(
-                function (ReflectionMethod $method) use (
-                    $modelName,
-                    $traitMethodNames,
-                    $currentMethod
-                ) {
-                    $methodName = $method->getName();
-                    /**
-                     * - The method must not originate in a trait
-                     * - It must not be a magic method
-                     * - It must be in the self scope and not inherited
-                     * - It must be in the this scope and not static
-                     * - It must not be an override of this one
-                     */
-                    if (!in_array($methodName, $traitMethodNames)
-                        && strpos($methodName, '__') !== 0
-                        && $method->class === $modelName
-                        && !$method->isStatic()
-                        && $methodName != $currentMethod
-                    ) {
-                        $parameters = (
-                            new ReflectionMethod($modelName, $methodName)
-                        )->getParameters();
-                        // If required parameters exist, this will be false and
-                        // omit this method
-                        return collect($parameters)->filter(
-                            function (ReflectionParameter $parameter) {
-                                // The method must have no required parameters
-                                return !$parameter->isOptional();
-                            }
-                        )->isEmpty();
-                    }
-                    return false;
+            ->map(function (ReflectionMethod $method) use ($filter, $model) {
+
+                if (!$method->getReturnType() || !$method->getReturnType()->getName()) {
+                    return null;
                 }
-            )
-            ->map(function (ReflectionMethod $method) use ($model, $filter) {
+
                 $methodName = $method->getName();
-                /** @var Relation|mixed $relation */
-                $relation = $model->$methodName();
+                $returnTypeClass = $method->getReturnType()->getName();
+
                 // Must return a Relation child. This is why we only want to do
                 // this once
-                if (is_subclass_of($relation, Relation::class)) {
-                    $type = (new ReflectionClass($relation))->getShortName();
+                if (is_subclass_of($returnTypeClass, Relation::class)) {
+                    $relation = $method->invoke($model);
```

Comment:
still not ideal but slightly safer than before. It at least checks the return typehint of the method before invoking it. Am also going to add a config to disable this by default and add a warning on the docs that this will invoke methods in the model to get the relation name.